### ISSUE TITLE
Out of memory issue with lru_maintainer on

### ISSUE THREAD

**Issue by panchenji** [memcached/memcached#195]
I tried to turned on slab_automove=1 in version 1.4.29. But it seems that the automove will only be executed when lru_maintainer is on. I checked the code and the logic has been move to lru_maintainer thread. But I did not see (maybe I missed) that it is mentioned in the protocol. So I turned on the lru_maintainer but I found tons of out of memory error when the memory is fully used. It seems that the lru tail eviction is not happening when there is not free chunks available. I did not see the issue when lru_maintainer is off. Could anyone help me take a look?

**Comment by dormando** [comment]
Can you include `stats settings` dump? You also need `-o slab_reassign` or it won't be able to move pages.

I'll take a closer look soon. That might be a glitch if automove is on but reassign is off (should be a startup error). Can you also try `-o modern` instead of specifying sub-options? Though you'll need 1.4.31 for that.

thanks

**Comment by panchenji** [comment]
I'm pretty sure I turned on the slab_ressign.

Here are the related logic for slab_automove I found in the code.
When I set slab_automove to 1, only the first code will be executed which stays in lru_maintainer_juggle function.
`    if (settings.slab_automove > 0 && chunks_free > (chunks_perslab * 2.5)) {
        slabs_reassign(slabs_clsid, SLAB_GLOBAL_PAGE_POOL);
    }`

`                    if (settings.slab_automove == 2) {
                        slabs_reassign(-1, orig_id);
                    }`

**Comment by dormando** [comment]
Could I get the `stats settings` output anyway? It would help a lot.

automove == 2 works out pretty badly in practice. it might get removed.

thanks!

**Comment by panchenji** [comment]
I have killed the process last week. But here is my -o params I used to start the process:  -o slab_reassign,slab_automove,maxconns_fast,lru_maintainer,hash_algorithm=murmur3

In my case, I want to use the slab_automove feature but not necessary lru_maintainer. But it would be great if lru_maintainer can give some performance improvement. 

I tested the slab move with mcperf and found that it only worked when lru_maintainer is on.
`./mcperf --num-conns=100 --conn-rate=1000 --sizes=0.01 --num-calls=10000 -s infra-cache-zen-dev-39c869c0 
 ./mcperf --num-conns=100 --conn-rate=1000 --sizes=0.001 --num-calls=10000 -s infra-cache-zen-dev-39c869c0
./mcperf --num-conns=100 --conn-rate=1000 --sizes=0.001 --num-calls=10000 -s infra-cache-zen-dev-39c869c0  -m get
`
Then I tested with -o slab_reassign,slab_automove,maxconns_fast,lru_maintainer,hash_algorithm=murmur3 with real traffic and tons of 
`SERVER_ERROR out of memory storing object` came into log when there is no free chunks any more.

**Comment by dormando** [comment]
Where'd you get mcperf from? I'll try to reproduce.

Enabling LRU maintainer does help memory efficiency in a lot of ways. It roughly breaks even CPU wise in my testing. It defers LRU shuffling to when the background thread gets to it, so fetches don't hold the LRU lock. It might look slightly worse in total CPU usage at the low end but scales much better with traffic and the latency is the same all along.

Either way, I'll take a look, thanks.

**Comment by panchenji** [comment]
You can find the mcperf here: https://github.com/twitter/twemperf

**Comment by dormando** [comment]
Are you using this with the default memory limits, btw? `-m 64`, you've only listed the -o related options.

thanks

**Comment by dormando** [comment]
Any chance I could get stats settings from your prod test? This problem isn't obvious yet.

To fully answer your initial question: Yes the slab automove only works with lru maintainer now. I'm not sure how clearly that's documented in the doc/protocol.txt file, and will double check.

The old automove code was a separate thread that very slowly made decisions... it was too slow to ever be effective, and LRU maintainer has a lot of information about what's going on so it works a lot better from there. There is still a hole in the automover if memory is full at all times in all slab classes. It does the movement decision by simply "reclaiming" pages from classes with enough free memory back into a global pool, then they get re-assigned when another class is under pressure.

So if every entry in your systems has a 0 TTL, it can't do anything. The LRU crawler (hopefully less intrusive after 1.4.31!) compliments the automover by clearing out expired memory, which then gets reclaimed to global, then etc. That's why `-o modern` is what it is.

That all said, with your test you shouldn't be getting OOM errors. It should simply cause evictions. Since you haven't shared the full startup options I can't tell if your memory limit is very low, if `-I` has been used to increase the memory limit, etc. When I fill memory with those startup options it just evicts a lot.

There are a few conditions which can cause OOM's, but I really need as much information as you can share. You may contact me privately if that helps as well.

thanks for your patience.

**Comment by panchenji** [comment]
For the automove issue, is there any case that user only need the automove but not the lru_maintainer feature? Or should we at least give an error when user only turn on slab_auotmove without lru_maintainer?


For the test environment,  here is my full command to start memcached. 
`./memcached -m 26500 -c 32768 -o slab_reassign,slab_automove,maxconns_fast,encoded_value,lru_maintainer,hash_algorithm=murmur3 -d`

I use 26.5G for the memory. But I remember the OOM issue also came out when I set it to 2.6G. 
Ideally, the eviction should happen when there is no free chunks. And it works well when I turn off the lru_maintainer. From the log I printed, it seems that for most of times, it failed to lock any item from the last of 5 of tail of COLD_LRU.

I dug into  a little bit more and did some code changes to find the issue. I think the main cause for the OOM is the new split LRU mechanism. If I always assign the new allocate item to COLD_LRU, turned off the COLD to WARM code for active item, and do unlink and link for item update. The issue is gone.

**Comment by dormando** [comment]
What's "encoded _value" in there?

1) automove without lru_maintainer would work for automove=2. I should make it a startup error under any other condition, yes.

2) I believe that you're seeing something but it doesn't happen under normal conditions, so it's definitely not related to simply having the lru_maintainer on or off. I can start up memcached with those options, flood it with 200 million sets, and it just does a ton of evictions, zero OOM's.

I'm not sure how your code change affected that. Do you have any other patches to the system, or is this just your debug code?

Do you have full stats output from after a run causes OOM's? It would be greatly helpful to get the full output of `stats`, `stats items`, `stats slabs` from an instance after it's been causing OOM's. there're a lot of stats I can comb through for information.

The "failed to lock item", means that the item is in active use for some reason. The weird thing is that evictions only ever happen from COLD_LRU in _both_ cases. With lru_maintainer items drop into COLD_LRU, then evict from the bottom. without lru_maintainer there's only a COLD_LRU, and items evict from that.

So simply moving items between HOT|WARM|COLD wouldn't cause item locks to be held at the _bottom_ of COLD_LRU for no reason. It'd be helpful to see the stats dump since that could help confirm what I'm suspecting.

thanks

**Comment by dormando** [comment]
(also: are you able to reproduce this at all with mcperf? or were you using mcperf to illustrate how automove worked? I haven't been able to repro yet).

**Comment by dormando** [comment]
just another general poke on this thread, thanks.

**Comment by panchenji** [comment]
You can ignore the encoded_value. I actually see the issue in the non-changed codes. 
I only use mcperf to check if automove works or not. 

For OOM issue, my traffic pattern is that it is with around 5200 qps get requests and around 1200 set requests. When you tested with 200 million set test, did you also do the get requests? Since without get traffic, eviction should always happen successfully. 

I did not find any logic issue from the code. I also believe ideally there should not be too much OOMs happened. But I feel that the new lru mechanism is less aggressive and take longer time to move a non-popular item to the end of cold lru.

**Comment by dormando** [comment]
Yes. Code releases are validated by running 12+ hour torture tests with mixed gets/sets totalling in the millions of ops/sec, while randomly shuffling pages between slab classes and running the LRU crawler. It doesn't cause OOM's.

Yes the background process is slightly less aggressive, but given 1200 sets/sec there's no way the background thread would get behind. You need to do 300,000+ generally. Even if it does get behind, it still advances each of the queues in a direct-reclaim mode.

I really honestly need the stats output I asked for before; it has a lot of counters related to reflocks/etc being found while trying to do evictions. I'm completely confident that this feature isn't broken as badly as it would seem; enabling lru maintainer doesn't just break evictions in all cases. There are regression tests, burn-in tests, and at minimum thousands of active users running with the feature.

You gave a hint that the items at the bottom of the LRU were locked for some reason, which is very difficult to cause. Either some (new?) code is leaking locks or something you're doing is holding locks. It's not clear if you meant reflocks or item mutex locks as well. Given your workload and how the mutex locks are used, I'd be very surprised if that were the case.

If items are reflocked, there are some defenses against that but that did change with the LRU maintainer. If reflocked items are frequently landing in your tail, it means your have client which are dying off or hanging without fully reading the data it requested (you can see clients which may be in this mode via `stats conns`).

The things I really need to know:
 * ideally full stats, stats items, stats slabs, dumps from a running instance which has exhibited the behavior. `stats conns` would be good too, but you should sanitize its output.
 * at very least the ratios of sets to OOM's and what slab classes they were happening on. as well as the lrutail_reflocked and lru_maintainer_juggles counters.

To rule out the more recent changes, you could try 1.4.28. I'm also not sure if you've tried 1.4.31 with these features enabled, as there have been some bugfixes since 1.4.29.

thanks!

**Comment by panchenji** [comment]
Here are the  stats you mentioned you the last post. 

[stats_conns.txt](https://github.com/memcached/memcached/files/457515/stats_conns.txt)
[stats_items.txt](https://github.com/memcached/memcached/files/457514/stats_items.txt)
[stats_slabs.txt](https://github.com/memcached/memcached/files/457513/stats_slabs.txt)
[stats.txt](https://github.com/memcached/memcached/files/457516/stats.txt)

**Comment by dormando** [comment]
Thank you! This helps narrow down what I need to test quite a lot.

**Comment by dormando** [comment]
Interesting:

```
STAT items:3:outofmemory 0
STAT items:4:outofmemory 21789845
STAT items:5:outofmemory 119629616
STAT items:6:outofmemory 16695361
STAT items:7:outofmemory 770911
STAT items:8:outofmemory 164951
STAT items:9:outofmemory 125178
STAT items:10:outofmemory 0
STAT items:11:outofmemory 0
STAT items:12:outofmemory 0
[rest are zero]
```

```
STAT items:3:moves_to_cold 0
STAT items:3:moves_to_warm 0
STAT items:4:moves_to_cold 0
STAT items:4:moves_to_warm 0
STAT items:5:moves_to_cold 0
STAT items:5:moves_to_warm 0
STAT items:6:moves_to_cold 0
STAT items:6:moves_to_warm 0
STAT items:7:moves_to_cold 23637766
STAT items:7:moves_to_warm 1972675
STAT items:8:moves_to_cold 17614437
STAT items:8:moves_to_warm 2263486
STAT items:9:moves_to_cold 12072967
STAT items:9:moves_to_warm 1953559
STAT items:10:moves_to_cold 12724624
STAT items:10:moves_to_warm 2345664
STAT items:11:moves_to_cold 10267559
STAT items:11:moves_to_warm 2629313
[rest are non-zero, similarly sized]
```

This is pretty odd. You were telling me that it never worked, but it looks like for low slab classes moves haven't happened (everything sitting in HOT)... however for higher classes the moves work fine and there are never OOM's. lrutail_reflocked isn't high enough to account for this (it would be > than the OOM errors).

Can you please explain what you did to see that was: "From the log I printed, it seems that for most of times, it failed to lock any item from the last of 5 of tail of COLD_LRU" ? exactly what lines did you inspect and what exactly did you see?

I don't believe it's an item lock since that wouldn't hang just some of the slab classes.

Also, any chance of seeing `stats settings` as well?

thanks!

**Comment by panchenji** [comment]
Here is the settings.
[stats_settings.txt](https://github.com/memcached/memcached/files/457820/stats_settings.txt)

Sorry for the unclear sentence before. "From the log I printed, it seems that for most of times, it failed to lock any item from the last of 5 of tail of COLD_LRU" I actually means that it failed to get the possible eviction item candidate in the lru_pull_tail function. I think in lru_pull_tail, it will try to lock any item in the tail and evict it if evict is set. I added a log after 
`        if ((hold_lock = item_trylock(hv)) == NULL)
            continue;`

But it did not print.

**Comment by dormando** [comment]
you mean like this?

```
        if ((hold_lock = item_trylock(hv)) == NULL)
            continue;
        printf("got this far\n");
```

**Comment by dormando** [comment]
For the instance the stats you gave me are from, are all of the slab classes used at all times, or only some some of the time?

Not sure how I can be super clear: Since the high slab classes have zero OOM's but a lot of moves, and the small slab classes had zero moves, but a lot of OOM's.. Were the small slab classes unused and then suddenly used, or is traffic uniform all the time?

**Comment by panchenji** [comment]
Yes. I added the log like that.

I believe all the slab classes are used all times and the traffic is uniform.

The weird thing is that for example slab 6, there is no free chunks left but there is no items moves between lrus.

**Comment by dormando** [comment]
Ok, thank you. Well the weird thing is that some slab classes work and some don't... but I'm a little lost as to how this could happen.

Need to think and run more tests. I might ask you to try a patch with some extra counters. I noticed the version listed isn't vanilla 1.4.29 for the instance you gave me. Since I've not seen your changes I have to go on faith that this isn't from a change on your part, is that safe to assume? I'm a little worried about spending a lot of time on this if it is.

Getting a lot closer though. thanks for your patience.

**Comment by panchenji** [comment]
This should be exactly the same codes as vanilla 1.4.29. I only added a default config file to make my packing script work. I did not change any code in this version.

**Comment by dormando** [comment]
I ask because in the version line:
`STAT version 1.4.29_1_gf332f28`

It would just be 1.4.29 if it were built from the tarball.

I'm looking into if the internal sizing of the sub-LRU's can get out of sync somehow. That would prevent anything from overflowing from HOT into COLD, and since evictions only come from COLD you'd see this scenario. I can't think of any other way it could happen.

Nothing obvious yet but I haven't had more than a few minutes to look into it. I'll add some asserts and run some load tests, but I might write a patch to expose the values so we can get a sample from your end. if I can't reproduce.

**Comment by panchenji** [comment]
I did some test and print some info and found an issue that might be the root cause. 

Please point it out if I'm wrong with anything.

In "do_item_alloc", the slab matching is based on total size of item (header + item). 
`    size_t ntotal = item_make_header(nkey + 1, flags, nbytes, suffix, &nsuffix);
    if (settings.use_cas) {
        ntotal += sizeof(uint64_t);
    }

    unsigned int id = slabs_clsid(ntotal);`

But in "lru_pull_tail", the limit comparison is based on sizes_bytes
`case WARM_LRU:
                if (limit == 0)
                    limit = total_bytes * settings.warm_lru_pct / 100;
                if (sizes_bytes[id] > limit) {`

And sizes_bytes(in do_item_link_q function) is the total size of all the items, but not counting for the header size. 

So let's say if our item size is 100 and the header size is 200, in the data I sent you before, it will be assigned to slab 6 which chunk size is 304. But the sizes_bytes of its hot lru will never exceed 33%. And no movement will happen inter-lru. That's also why we only see this issue in smaller size slab.

From the data I print in my test, we definitely have header which has size over 200. I think it will also be the same problem for other users. 

To fix this issue, we should probably count the total size in sizes_bytes rather than item only size.

**Comment by dormando** [comment]
Was suspicious about that last night. Need to triple confirm and run more tests, but it does look like I intended to do ITEM_ntotal() instead of it->nbytes. Previous to 1.4.29 this would just count the number of items, but I had to switch it so large items could be proportioned correctly.

Would've been easier to just keep counts of active chunks but I wasn't able to make that working before release...

Thanks for sticking with this. Being able to finally see the full stats output made the area of interest obvious and made the rest easy to do, if this indeed the bug :)

**Comment by dormando** [comment]
https://github.com/dormando/memcached/tree/sizes_bytes_fix can you try this tree, please?

I need to head out shortly but this passes tests. Need to spend some more time with it and add proper regression tests and re-audit for any similar bugs.. At a glance it does look like other parts of the new cod uses ITEM_ntotal() properly.

In the meantime it would be helpful to see if this fixes your problem.

**Comment by panchenji** [comment]
I changed to compare the item number in each lru (if hot lru use more than 33% of chunks, move it) and it worked pretty well. I'll test your fix soon.

**Comment by ljagiello** [comment]
Hi,

I think I have the same problem as described here:

Current configuration:
```
memcached -u nobody -m88064 -c9216 -p11000 -n 5 -f 1.05 -o modern,slab_chunk_max=8192
```

Memcached stats:
- [stats.conns](https://dl.dropboxusercontent.com/u/133846/stats.conns)
- [stats.items](https://dl.dropboxusercontent.com/u/133846/stats.items)
- [stats.settings](https://dl.dropboxusercontent.com/u/133846/stats.settings)
- [stats.slabs](https://dl.dropboxusercontent.com/u/133846/stats.slabs)

Eviction graph - https://www.dropbox.com/s/yjfv0wh2eupaveo/Screenshot%202016-09-07%2015.19.49.png?dl=0 almost instantly when we see evictions application starts reporting out of memory errors.

It started with upgrade to 1.4.29 (previously it was some default ubuntu version 1.4.14), as soon as cache was full we started noticing errors like: `MemcachedClient: store failed - SERVER_ERROR out of memory storing object`. After upgrade to 1.4.31 the problem is still the same, as soon as cache is full we have problems with out of memory.

**Comment by dormando** [comment]
@ljagiello any chance you could try this branch as well?

**Comment by ljagiello** [comment]
@dormando sure I will try that branch.

**Comment by dormando** [comment]
Either of you have any luck with the patch?

**Comment by ljagiello** [comment]
@dormando so far it's running I need ~2 weeks before my cache will be full, I will try on weekend to reproduce the issue in some controlled environment.

**Comment by panchenji** [comment]
I have tested it through the weekend and it works well.

**Comment by dormando** [comment]
thank you!

**Comment by senugula** [comment]
@dormando we have tested it over past 3 days it works well, we don't see no memory , When do plan of releasing with this fix? Thank you !

**Comment by ljagiello** [comment]
Same here, looks stable.

**Comment by dormando** [comment]
Very soon! Was sick last week, apologies

**Comment by dormando** [comment]
released now in 1.4.32, sorry for the very long delay :(
